splice_machine/src/main/java/com/splicemachine/derby/impl/sql/compile/NestedLoopJoinStrategy.java

```diff
@@ -319,14 +321,17 @@ private double getNljOnSparkPenalty(Optimizable table,
         double multiplier = innerCost.getFromBaseTableRows();
         if (multiplier < 1d)
             multiplier = 1d;
-        // TODO:  Enable this code when DB-11521 is fixed.
-        //        Currently join cardinality is grossly underestimated,
-        //        so what we think is a safe nested loop join may in fact
-        //        be joining hundreds or thousands of rows from the left table.
-//        if (!isBaseTable(table))
-//            return NLJ_ON_SPARK_PENALTY * multiplier;
-//        if (hasJoinPredicateWithIndexKeyLookup(predList) && outerCost.rowCount() <= 10)
-//            return retval;
+        QueryTreeNode queryTreeNode = (QueryTreeNode) table;
+        LanguageConnectionContext lcc = queryTreeNode.getLanguageConnectionContext();
+        Boolean olapAlwaysPenalizeNLJ = (Boolean)
+            lcc.getSessionProperties().getProperty(SessionProperties.PROPERTYNAME.OLAPALWAYSPENALIZENLJ);
+
+        if (olapAlwaysPenalizeNLJ == null || !olapAlwaysPenalizeNLJ.booleanValue()) {
+            if (!isBaseTable(table))
```

Comment:
@msirek Should we allow IndexLookups here too? We could also have ProjectRestricts before the Scan.

Reply:
@dgomezferro IndexLookups are built automatically.  A FromBaseTable node still describes the access.  I have added support for ProjectRestricts on top of the scan.